docs/architecture/adr-065-store-v2.md

```diff
@@ -0,0 +1,259 @@
+as research and benchmarking continues. However, we propose to hardcode the use
+of RocksDB as the primary physical storage backend.
+
+Since the SS layer will be implemented as a `KVStore`, it will support the
+following functionality:
+
+* Range queries
+* CRUD operations
+* Historical queries and versioning
+* Pruning
+
+The RMS will keep track of all buffered writes using a dedicated and internal
+`MemoryListener` for each `StoreKey`. For each block height, upon `Commit`, the
+SS layer will write all buffered (key, value) pairs under a [RocksDB user-defined timestamp](https://github.com/facebook/rocksdb/wiki/User-defined-Timestamp-%28Experimental%29) column
+family using the block height as the timestamp, which is an unsigned integer.
+This will allow a client to fetch (key, value) pairs at historical and current
+heights along with making iteration and range queries relatively performant as
+the timestamp is the key suffix.
+
+Note, we choose not to use a more general approach of allowing any embedded key/value
+database, such as LevelDB or PebbleDB, using height key-prefixed keys to
+effectively version state because most of these databases use variable length
+keys which would effectively make actions likes iteration and range queries less
+performant.
+
+The state machine will continue to query state through it's own dedicated `KVStore`
+using the relevant `KVStoreKey`. However, the query will be invoked against the
+SS layer. Client queries, on the other hand, will be routed through
+the `MultiStore` set in `BaseApp` via a `abci.RequestQuery`. This in effect will
+route the query to the RMS and the RMS will use the SS layer to perform a direct
+key lookup. If a version is specified the `abci.RequestQuery` object, the SS `KVStore`
+will set the appropriate value for the timestamp. If no height is provided, the
+SS layer will assume the latest height. The SS layer will store a reverse index
+to lookup `LatestVersion -> timestamp(version)` which is set on `Commit`.
+
+Since operators might want pruning strategies to differ in SS compared to SC,
+e.g. having a very tight pruning strategy in SC while having a looser pruning
+strategy for SS, we propose to introduce an additional pruning configuration,
+with parameters that are identical to what exists in the SDK today, and allow
+operators to control the pruning strategy of the SS layer independently of the
+SC layer.
+
+#### Proofs
+
+Since the SS layer is naturally a storage layer only, without any commitments
+to (key, value) pairs, it cannot provide Merkle proofs to clients during queries.
+
+Since the pruning strategy against the SC layer is configured by the operator,
+we can therefore have the RMS route the query SC layer if the version exists and
+`prove: true`. Otherwise, the query will fall back to the SS layer without a proof.
+
+We could explore the idea of using state snapshots to rebuild an in-memory IAVL
```

Comment:
I thought we decided to go this path for historical versions. Why are we saying we could?